tempto-core/src/main/java/com/teradata/tempto/internal/hadoop/hdfs/SpnegoWebHdfsClient.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.teradata.tempto.internal.hadoop.hdfs;
+
+import com.google.common.base.Throwables;
+import com.google.common.net.HostAndPort;
+import com.teradata.tempto.kerberos.KerberosAuthentication;
+import org.apache.http.auth.AuthSchemeProvider;
+import org.apache.http.auth.AuthScope;
+import org.apache.http.auth.Credentials;
+import org.apache.http.client.config.AuthSchemes;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.config.Lookup;
+import org.apache.http.config.RegistryBuilder;
+import org.apache.http.impl.auth.SPNegoSchemeFactory;
+import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.DefaultHttpRequestRetryHandler;
+import org.apache.http.impl.client.HttpClientBuilder;
+
+import javax.security.auth.Subject;
+
+import java.io.IOException;
+import java.security.Principal;
+import java.security.PrivilegedAction;
+
+public class SpnegoWebHdfsClient
+        extends AbstractWebHdfsClient
+{
+
+    private final KerberosAuthentication kerberosAuthentication;
+    private final CloseableHttpClient httpClient;
+
+    public SpnegoWebHdfsClient(HostAndPort nameNode, String username, String keytab)
+    {
+        super(nameNode, username);
+        this.kerberosAuthentication = new KerberosAuthentication(username, keytab);
+        this.httpClient = createSpnegoHttpClient();
+    }
+
+    private CloseableHttpClient createSpnegoHttpClient()
+    {
+        HttpClientBuilder builder = HttpClientBuilder.create();
+        Lookup<AuthSchemeProvider> authSchemeRegistry = RegistryBuilder.<AuthSchemeProvider>create()
+                .register(AuthSchemes.SPNEGO, new SPNegoSchemeFactory(true)).build();
+        builder.setDefaultAuthSchemeRegistry(authSchemeRegistry);
+        BasicCredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+        credentialsProvider.setCredentials(new AuthScope(null, -1, null), new NullCredentials());
+        builder.setDefaultCredentialsProvider(credentialsProvider);
+        builder.setRetryHandler(new DefaultHttpRequestRetryHandler(3, true));
+        return builder.build();
+    }
+
+    @Override
+    protected CloseableHttpResponse executeRequest(HttpUriRequest request)
```

Comment:
That is what i did at the beggining. Unfortunately the way we execute HTTP requests is not the only difference between the authentications. In case of simple authentication we must specify user as the `user.name` query parameter, so, it must be considered during the HTTP request creation. Http request is immutable, so, in order to change URL we must wrap it somehow. I will try to do that.